HCore-Web/Providers/Impl/DownloadProcessingProxyUrlProviderImpl.cs

```diff
@@ -0,0 +1,197 @@
+                    {
+                        mimeType = response.Content.Headers.ContentType.MediaType;
+                    }
+
+                    characterSet = response.Content.Headers.ContentType.CharSet;
+                }
+                else
+                {
+                    _logger.LogError(
+                        $"Failed to download file from source location {downloadUri}. " +
+                        $"file name: {JsonConvert.SerializeObject(fileName)}. " +
+                        $"HTTP result was: {response.StatusCode} {response.ReasonPhrase}, response body: {response.Content}"
+                    );
+
+                    throw new HttpRequestException("Failed to download file from source location!");
+                }
+            }
+
+            return new DownloadFileDataImpl()
+            {
+                data = fileData,
+                FileName = fileName,
+                CharacterSet = characterSet,
+                MimeType = mimeType,
+            };
+        }
+
+        public virtual Uri GetSourceUri(HttpRequest request)
+        {
+            return new Uri(request.Query["u"]);
+        }
+
+        private dynamic ConvertParametersFromQueryData(HttpRequest request)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (!request.Query.TryGetValue(UrlQueryKeyName, out var queryData) || queryData.Count != 1)
+            {
+                return null;
+            }
+
+            string protectedJson = queryData[0];
+            string json = _protector.Unprotect(protectedJson);
+            return DeserializeObject<dynamic>(json);
+        }
+
+        private string ConvertParametersToQueryData(dynamic downloadParameters)
+        {
+            string json = SerializeObject(downloadParameters);
+            string protectedJson = _protector.Protect(json);
+            return $"{UrlQueryKeyName}={Uri.EscapeDataString(protectedJson)}" ;
+        }
+
+        private string CalculateHashFromParameters(Uri downloadSourceUri, string fileName, string downloadMimeType = null)
+        {
+            string valueToHash = $"{fileName}:{downloadMimeType}:{downloadSourceUri}";
+
+            // see https://stackoverflow.com/questions/33245247/hashalgorithms-in-coreclr
+            using var algorithm = SHA256.Create();
+
+            // Create the at_hash using the access token returned by CreateAccessTokenAsync.
+            var hash = algorithm.ComputeHash(Encoding.UTF8.GetBytes(valueToHash));
+
+            return ToBase64String(hash);
+        }
+
+        private bool IsHashValid(string hashToVerify, Uri downloadSourceUri, string fileName, string downloadMimeType = null)
+        {
+            string calculatedHash = CalculateHashFromParameters(downloadSourceUri, fileName, downloadMimeType);
+            return calculatedHash != null && calculatedHash.Equals(hashToVerify);
```

Comment:
Please use string.Equals without null check

Reply:
Done. Seems I am too used to Java, which is lacking this useful function